File: Android/app/src/main/java/com/hhp227/paging3/paging/LazyPagingItems.kt
package com.hhp227.paging3.paging

import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import androidx.compose.foundation.lazy.LazyItemScope
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.runtime.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.filterNotNull
import kotlinx.coroutines.flow.onStart

class LazyPagingItems<T : Any> internal constructor(
    private val flow: Flow<PagingData<T>>
) : DifferCallback {
    var itemSnapshotList by mutableStateOf(
        ItemSnapshotList<T>(0, 0, emptyList())
    )
        private set

    val itemCount: Int get() = pagingDataDiffer.size

    private val pagingDataDiffer = PagingDataDiffer<T>(
        differCallback = this,
        updateItemSnapshotList = ::updateItemSnapshotList
    )

    private fun updateItemSnapshotList() {
        itemSnapshotList = pagingDataDiffer.snapshot()
    }

    operator fun get(index: Int): T? {
        pagingDataDiffer[index] // this registers the value load
        return itemSnapshotList[index]
    }

    fun peek(index: Int): T? {
        return itemSnapshotList[index]
    }

    fun retry() {
        pagingDataDiffer.retry()
    }

    fun refresh() {
        pagingDataDiffer.refresh()
    }

    var loadState: CombinedLoadStates by mutableStateOf(
        CombinedLoadStates(
            refresh = InitialLoadStates.refresh,
            prepend = InitialLoadStates.prepend,
            append = InitialLoadStates.append,
            source = InitialLoadStates
        )
    )
        private set

    internal suspend fun collectLoadState() {
        pagingDataDiffer.loadStateFlow.filterNotNull().collect {
            loadState = it
        }
    }

    internal suspend fun collectPagingData() {
        flow.onStart { Log.e("IDIS_TEST", "TEST3") }.collectLatest(pagingDataDiffer::collectFrom)
    }

    override fun onChanged(position: Int, count: Int) {
        if (count > 0) {
            updateItemSnapshotList()
        }
    }

    override fun onInserted(position: Int, count: Int) {
        if (count > 0) {
            updateItemSnapshotList()
        }
    }

    override fun onRemoved(position: Int, count: Int) {
        if (count > 0) {
            updateItemSnapshotList()
        }
    }
}

private val InitialLoadStates = LoadStates(
    LoadState.Loading,
    LoadState.NotLoading(false),
    LoadState.NotLoading(false)
)

@Composable
fun <T : Any> Flow<PagingData<T>>.collectAsLazyPagingItems(): LazyPagingItems<T> {
    val lazyPagingItems = remember(this) { LazyPagingItems(this) }

    LaunchedEffect(lazyPagingItems) {
        lazyPagingItems.collectPagingData()
    }
    LaunchedEffect(lazyPagingItems) {
        lazyPagingItems.collectLoadState()
    }
    return lazyPagingItems
}

fun <T : Any> LazyListScope.items(
    items: LazyPagingItems<T>,
    key: ((item: T) -> Any)? = null,
    itemContent: @Composable LazyItemScope.(value: T?) -> Unit
) {
    Log.e("TEST", "??? ${items.itemCount}")
    items(
        count = items.itemCount,
        key = if (key == null) null else { index ->
            val item = items.peek(index)
            if (item == null) {
                PagingPlaceholderKey(index)
            } else {
                key(item)
            }
        }
    ) { index ->
        itemContent(items[index])
    }
}

fun <T : Any> LazyListScope.itemsIndexed(
    items: LazyPagingItems<T>,
    key: ((index: Int, item: T) -> Any)? = null,
    itemContent: @Composable LazyItemScope.(index: Int, value: T?) -> Unit
) {
    items(
        count = items.itemCount,
        key = if (key == null) null else { index ->
            val item = items.peek(index)
            if (item == null) {
                PagingPlaceholderKey(index)
            } else {
                key(index, item)
            }
        }
    ) { index ->
        itemContent(index, items[index])
    }
}

private data class PagingPlaceholderKey(private val index: Int) : Parcelable {
    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(index)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object {
        @Suppress("unused")
        @JvmField
        val CREATOR: Parcelable.Creator<PagingPlaceholderKey> =
            object : Parcelable.Creator<PagingPlaceholderKey> {
                override fun createFromParcel(parcel: Parcel) =
                    PagingPlaceholderKey(parcel.readInt())

                override fun newArray(size: Int) = arrayOfNulls<PagingPlaceholderKey?>(size)
            }
    }
}

File: iOS/Paging/Paging/LazyPagingItems.swift
//
//  LazyPagingItems.swift
//  Paging
//
//  Created by 홍희표 on 2022/06/19.
//

import Foundation
import Combine
import SwiftUI

// TODO 버그내용: 하단에 가면 Insert가 너무 많이 호출됨
public class LazyPagingItems<T : Any>: ObservableObject, DifferCallback {
    private let publisher: any Publisher<PagingData<T>, Never>
    
    @Published private(set) var itemSnapshotList = ItemSnapshotList<T>(0, 0, [])

    var itemCount: Int {
        get {
            return pagingDataDiffer.size
        }
    }
    
    private lazy var pagingDataDiffer: PagingDataDiffer<T> = PagingDataDiffer<T>(
        differCallback: self,
        updateItemSnapshotList: updateItemSnapshotList
    )

    private var subscriptions = Set<AnyCancellable>()

    private func updateItemSnapshotList() {
        DispatchQueue.main.async { // TODO 지울것
            self.itemSnapshotList = self.pagingDataDiffer.snapshot()
        }
        
    }

    func get(_ index: Int) -> T? {
        let _ = pagingDataDiffer.get(index: index)
        return itemSnapshotList[index]
    }

    func peek(_ index: Int) -> T? {
        return itemSnapshotList[index]
    }

    func retry() {
        pagingDataDiffer.retry()
    }

    func refresh() {
        pagingDataDiffer.refresh()
    }

    @Published private(set) var loadState: CombinedLoadStates = CombinedLoadStates(
        refresh: InitialLoadStates.refresh,
        prepend: InitialLoadStates.prepend,
        append: InitialLoadStates.append,
        source: InitialLoadStates
    )
    
    internal func collectLoadState() {
        pagingDataDiffer.loadStatePublisher.compactMap { $0 }.assign(to: \.loadState, on: self).store(in: &subscriptions)
    }
    
    internal func collectPagingData() {
        publisher.eraseToAnyPublisher()
            .buffer(size: 1, prefetch: .keepFull, whenFull: .dropOldest)
            .sink(receiveValue: self.pagingDataDiffer.collectFrom)
            .store(in: &subscriptions)
    }

    func onChanged(_ position: Int, _ count: Int) {
        if count > 0 {
            updateItemSnapshotList()
        }
    }

    func onInserted(_ position: Int, _ count: Int) {
        if count > 0 {
            updateItemSnapshotList()
        }
    }

    func onRemoved(_ position: Int, _ count: Int) {
        if count > 0 {
            updateItemSnapshotList()
        }
    }

    init(_ publisher: any Publisher<PagingData<T>, Never>) {
        self.publisher = publisher
    }
}

private let InitialLoadStates = LoadStates(
    refresh: .Loading.instance,
    prepend: .NotLoading(false),
    append: .NotLoading(false)
)

extension Publisher where Failure == Never {
    func collectAsLazyPagingItems<T>() -> LazyPagingItems<T> where Output == PagingData<T> {
        @State var lazyPagingItems = LazyPagingItems(self)

        DispatchQueue.main.async {
            lazyPagingItems.collectPagingData()
        }
        DispatchQueue.main.async {
            lazyPagingItems.collectLoadState()
        }
        return lazyPagingItems
    }
}

extension ForEach where Data == Range<Int>, ID == Int, Content : View {
    public init<T>(_ data: LazyPagingItems<T>, @ViewBuilder content: @escaping (T?) -> Content) {
        print("???? \(data.itemCount)")
        self.init(0..<data.itemCount, id: \.self) { index in
            content(data.get(index))
        }
    }
}

/*extension ForEach where Data == Range<Int>, ID == Int, Content == AnyView {
    public init<T>(_ data: LazyPagingItems<T>, @ViewBuilder content: @escaping (T?) -> Content) {
        self.init(0..<data.itemCount, id: \.self) { index in
            AnyView(content(data.get(index)).onAppear {
                data.get(index)
            })
            // TODO 여기서. onAppear로 accessHint접근하는방법 알아보기
        }
    }
}*/
